### ISSUE TITLE
Typst Writer: Proper captions for tables

### ISSUE THREAD

**Issue by iandol** [jgm/pandoc#9194]
At present, the Typst writer exports a markdown table like this:

```markdown
|       |  A   |    B   |    C    |    D    |
|-------|-----:|-------:|--------:|--------:|
| Ratio | 0.89 |   0.93 |    1.11 |     0.9 |
: Some Ratios
```

```typst
#align(center)[#table(
  columns: 5,
  align: (col, row) => (auto,right,right,right,right,).at(col),
  inset: 6pt,
  [], [A], [B], [C], [D],
  [Ratio],
  [0.89],
  [0.93],
  [1.11],
  [0.9],
)
#align(center, [Some Ratios])
]
```

The caption is treated as text and just placed below the table. However, the "official" method is to wrap the table in a figure:

```typst
#figure(
  table(
    columns: 4,
    [t], [1], [2], [3],
    [y], [0.3s], [0.4s], [0.8s],
  ),
  caption: [Timing results],
)
```
https://typst.app/docs/reference/meta/figure/

![](https://typst.app/assets/docs/TruCOIZS9t5lQvh9KKRu9gAAAAAAAAAA.png)

This allows labels with counters and can then also be cross-referenced elsewhere...

Looking at https://github.com/jgm/pandoc/blob/main/src/Text/Pandoc/Writers/Typst.hs#L158 I could make a pull request though I know no haskell, so would just guesstimate which bits to change…

**Pandoc version?**
3.1.9 macOS

**Comment by iandol** [comment]
Note that Quarto also rewrites Pandoc's output to emit tables wrapped in figures, perhaps @cscheid from the Quarto team has more details to share:

```typst
#figure([
#align(center)[#table(
  columns: 5,
  align: (col, row) => (auto,right,right,right,right,).at(col),
  inset: 6pt,
  [], [Lion], [Rabbit], [Tiger], [Horse],
  [Ratio],
  [0.89],
  [0.93],
  [1.11],
  [0.9],
)
]

], caption: figure.caption(
position: top, 
[
Caption.
]), 
kind: "quarto-float-tbl", 
supplement: "Table", 
numbering: "1", 
)
```

**Comment by cscheid** [comment]
> perhaps @cscheid from the Quarto team has more details to share:

Pandoc probably doesn't want to do anything special for Quarto here, since our crossref system is significantly more opinionated.

I'll just note here that `#figure`, in Typst parlance, is closer to a "float" in LaTeX. (Notice the `kind` field that controls which counter to use, etc). A pure Typst cross-referenceable element will also need a label associated with it, but I see that [src/Text/Pandoc/Writers/Typst.hs](https://github.com/jgm/pandoc/commit/3591733e19f3a8db200c6170e6e543eb97e91728#diff-ea339672c196cff4a46d8707aba5fc9c6ccc37b2ddf700ee6d27415e96ad10e3) will add the label based on the table identifier, so this should work well.

Typst will also automatically use a Table counter for "floats" with Table elements, so the fix from jgm in 3591733 is precisely what I'd do in Pandoc.

**Comment by iandol** [comment]
I can confirm I get captioned tables with numbering in Typst outputs with the latest Pandoc nightly, thank you @jgm! And thank you @cscheid for your input.
